sklearn/_fast_gradient_boosting/histogram.pyx

```diff
@@ -0,0 +1,459 @@
+        const G_H_DTYPE_C [::1] ordered_gradients,  # IN
+        const G_H_DTYPE_C [::1] ordered_hessians,  # IN
+        hist_struct [:, ::1] out) nogil:  # OUT
+    """Return histogram for a given feature."""
+    cdef:
+        unsigned int i = 0
+        unsigned int n_node_samples = sample_indices.shape[0]
+        unsigned int unrolled_upper = (n_node_samples // 4) * 4
+
+        unsigned int bin_0
+        unsigned int bin_1
+        unsigned int bin_2
+        unsigned int bin_3
+        unsigned int bin_idx
+
+    for i in range(0, unrolled_upper, 4):
+        bin_0 = binned_feature[sample_indices[i]]
+        bin_1 = binned_feature[sample_indices[i + 1]]
+        bin_2 = binned_feature[sample_indices[i + 2]]
+        bin_3 = binned_feature[sample_indices[i + 3]]
+
+        out[feature_idx, bin_0].sum_gradients += ordered_gradients[i]
+        out[feature_idx, bin_1].sum_gradients += ordered_gradients[i + 1]
+        out[feature_idx, bin_2].sum_gradients += ordered_gradients[i + 2]
+        out[feature_idx, bin_3].sum_gradients += ordered_gradients[i + 3]
+
+        out[feature_idx, bin_0].sum_hessians += ordered_hessians[i]
+        out[feature_idx, bin_1].sum_hessians += ordered_hessians[i + 1]
+        out[feature_idx, bin_2].sum_hessians += ordered_hessians[i + 2]
+        out[feature_idx, bin_3].sum_hessians += ordered_hessians[i + 3]
+
+        out[feature_idx, bin_0].count += 1
+        out[feature_idx, bin_1].count += 1
+        out[feature_idx, bin_2].count += 1
+        out[feature_idx, bin_3].count += 1
+
+    for i in range(unrolled_upper, n_node_samples):
+        bin_idx = binned_feature[sample_indices[i]]
+        out[feature_idx, bin_idx].sum_gradients += ordered_gradients[i]
+        out[feature_idx, bin_idx].sum_hessians += ordered_hessians[i]
+        out[feature_idx, bin_idx].count += 1
+
+
+cpdef void _build_histogram_no_hessian(
+        const int feature_idx,
+        const unsigned int [::1] sample_indices,  # IN
+        const X_BINNED_DTYPE_C [::1] binned_feature,  # IN
+        const G_H_DTYPE_C [::1] ordered_gradients,  # IN
+        hist_struct [:, ::1] out) nogil:  # OUT
+    """Return histogram for a given feature, not updating hessians.
+
+    Used when the hessians of the loss are constant (typically LS loss).
+    """
+
+    cdef:
+        unsigned int i = 0
+        unsigned int n_node_samples = sample_indices.shape[0]
+        unsigned int unrolled_upper = (n_node_samples // 4) * 4
+
+        unsigned int bin_0
+        unsigned int bin_1
+        unsigned int bin_2
+        unsigned int bin_3
+        unsigned int bin_idx
+
+    for i in range(0, unrolled_upper, 4):
+        bin_0 = binned_feature[sample_indices[i]]
+        bin_1 = binned_feature[sample_indices[i + 1]]
+        bin_2 = binned_feature[sample_indices[i + 2]]
+        bin_3 = binned_feature[sample_indices[i + 3]]
+
+        out[feature_idx, bin_0].sum_gradients += ordered_gradients[i]
```

Comment:
This is to optimize auto-vectorization. It helps gcc :)
@NicolasHug it would be interesting to try not to manually unroll the loop and compile with the `-Ofast` flag. If performance and numerical stability are preserved, I think it would be worth, and the code would be simpler.

Reply:
`-Ofast` enables all `-O3` optimizations, which are known to produce really buggy code. I'd rather not do that. It'll be fine to enable the specific optimizations which you have in mind though.

Reply:
`-Ofast` enables way more than `-O3` :)
In fact, `-O3` is enabled by default when building sklearn. If you look at the build log, you'll see `-O3` all over the place. This flag is safe. It should produce correct code w.r.t. floating point arithmetic.

`-Ofast` on the other hand enables unsafe math operations, e.g. make you of associativity rules.
However, in this file there's nothing to optimize this way so I think it should be safe.

Reply:
I'm not sure how I feel either about using `Ofast`. If we re-wrap this code and use `Ofast` but in a future decide not to use `Ofast` anymore for whatever reason, the code will be unnecessarily slow.

I agree however that this unwarpping is a bit unsettling when you first encounter it. I'll add a note at the beginning of the file to explain it.

Reply:
I want to add that scipy enables the `-ffast-math` flag (which is enabled by `-Ofast`), which enables the `-funsafe_math_optimizations` flag. Since sklearn makes intensive use of scipy, we sort of already effectively do it in sklearn :)